Add/crud get total count selector

louwie17 opened the pull request:
### All Submissions:

-   [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md)?
-   [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
-   [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change?

<!-- Mark completed items with an [x] -->

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

Add's a getItemsTotalCount selector to the crud data store, so we can use it for the product categories data store.
This is required for the category dropdown field.

### How to test the changes in this Pull Request:

1. Load this branch, finish the onboarding and create some Product categories ( an easy way is also by importing the sample products in the **Personalize your store** task, which adds categories by default.
2. Go to **WooCommerce > Home** and open your console.
3. Retrieve a normal list of product categories `wp.data.select('experimental/wc/admin/products/categories').getProductCategories({ per_page: 5 })`
4. Test the total count selector: `wp.data.select('experimental/wc/admin/products/categories').getProductCategoriesTotalCount({ per_page: 5 })` this should return the total count, without triggering another request.
5. Test the total count selector: `wp.data.select('experimental/wc/admin/products/categories').getProductCategoriesTotalCount({ parent: 1 })` this should return `undefined` first and then the total count a second time, as it does trigger a new request.
6. Run `wp.data.select('experimental/wc/admin/products/categories').getProductCategories({ per_page: 5 })` one more time and look at the category list, and get one of the category id's
7. Run `wp.data.select('experimental/wc/admin/products/categories').getProductCategory( 222 )`, replacing 222 with the category id. It should return the category without triggering another request.
If you refresh the page and only run this, it should trigger a request and return the category when calling it a second time.


### Other information:

-   [x] Have you added an explanation of what your changes do and why you'd like us to include them?
-   [x] Have you written new tests for your changes, as applicable?
-   [x] Have you successfully run tests with your changes locally?
-   [x] Have you created a changelog file for each project being changed, ie `pnpm changelog add --filter=<project>`?

<!-- Mark completed items with an [x] -->

### FOR PR REVIEWER ONLY:

-   [ ] I have reviewed that everything is sanitized/escaped appropriately for any SQL or XSS injection possibilities. I made sure Linting is not ignored or disabled.

louwie17 committed:
Add getItemsTotalCount to crud data store

louwie17 committed:
Add changelog

louwie17 committed:
Make getTotalItemsCount more generic

mattsherman commented on packages/js/data/src/crud/reducer.ts:
```diff
@@ -48,6 +51,17 @@ export const createReducer = () => {
+						...state,
+						itemsCount: {
+							...state.itemsCount,
+							[ getTotalOrderCountResourceName(
```
It makes no sense to me why a general function is calling something that specifically references "Order". Either I'm misinterpreting things or `getTotalOrderCountResourceName` should be renamed.

louwie17 replied:
Aaah yes, I missed that from my copy paste (facepalm)

mattsherman commented on packages/js/data/src/crud/selectors.ts:
```diff
@@ -113,6 +113,20 @@ export const getItems = createSelector(
+	state: ResourceState,
+	query: ItemQuery | string
+) => {
+	const itemQuery = getTotalOrderCountResourceName(
```
Same comment as above... why the reference to "Order" in here?

mattsherman commented on packages/js/data/src/utils.ts:
```diff
@@ -21,6 +21,25 @@ export function getResourceName(
+ * @param {Object} query  Query for order totals count.
+ * @return {string} Resource name for order totals.
+ */
+export function getTotalOrderCountResourceName(
```
This feels misnamed (see related comments above)... maybe `getTotalCountResourceName` would be better?

mattsherman commented on packages/js/data/src/utils.ts:
```diff
@@ -21,6 +21,25 @@ export function getResourceName(
+	prefix: string,
+	query: Record< string, unknown >
+) {
+	const { _fields, page, per_page, ...totalsQuery } = query;
```
Is there a more robust way to handle this? It seems like this is brittle, requiring updating in the future if additional query params are added that don't affect totals. Perhaps we should explicitly list what query params we do want, instead of which ones we don't want?

What query params do we want? Are there any?

louwie17 replied:
> What query params do we want? Are there any?

The problem with the params we do want is that they can change depending on what product you are querying for. For the product categories we will want the `parent`, `id`, `menu_order`, `count`, `slug`, `display`. But this might be different for just `products` or attribute terms.
`_fields`, `page`, and `per_page` are generic across all the REST api GET requests, and are currently the only items that don't affect the total of number of items being returned.

mattsherman replied:
Makes sense.

mattsherman reviewed (changes requested):
> The getItemsTotalCount does not have a resolver and relies on the getProductCategories being called.

This behavior feels wrong to me. I would expect to be able to call `getItemsTotalCount` without calling anything else previously. Could we change it to work this way? There might be scenarios where the total counts are needed solely, or at least before any individual items are queried for.

I also left a comment about the `getTotalOrderCountrResourceName` feeling misnamed.

louwie17 commented:
> > The getItemsTotalCount does not have a resolver and relies on the getProductCategories being called.
> 
> This behavior feels wrong to me. I would expect to be able to call `getItemsTotalCount` without calling anything else previously. Could we change it to work this way? There might be scenarios where the total counts are needed solely, or at least before any individual items are queried for.

Thanks for the review @mattsherman, and yeah I do agree with you here. I just found it a bit annoying that this does mean it will always send an additional request, even if `getItems` has already been triggered and the total count been set. I will double check if there is a way around this or not.

louwie17 committed:
Rename total count resource name function, and add resolver for getTotalCount

louwie17 commented:
@mattsherman I addressed your changes, also in relation to below:

> The getItemsTotalCount does not have a resolver and relies on the getProductCategories being called.

I added a resolver for the `getItemsTotalCount`, and to prevent a double request from happening I was able to trigger `finishResolution` from within the `getItems` resolver. This seems to work well.
I followed an [example from WCPay](https://github.com/Automattic/woocommerce-payments/blob/4697ea36523a1175582a0d9519a85bd7e180fa6a/client/data/deposits/resolvers.js#L133-L136) and also added the `finishResolution` for the `getItem` request for each individual item.
I will update the testing instructions to reflect this.

mattsherman commented on packages/js/data/src/crud/resolvers.ts:
```diff
@@ -1,7 +1,7 @@
  * External dependencies
  */
-import { apiFetch } from '@wordpress/data-controls';
+import { apiFetch, dispatch } from '@wordpress/data-controls';
```
`dispatch` control in `@wordpress/data-controls` is deprecated. You should use the `dispatch` control in `@wordpress/data` instead.

See: https://github.com/WordPress/gutenberg/blob/b641a05e3d50f1ffe186543280510871bfdfe459/packages/data-controls/src/index.js#L70

louwie17 replied:
Good catch, let me update that

louwie17 replied:
Updated as part of: https://github.com/woocommerce/woocommerce/pull/34613/commits/6c84803aefdf5c1e63e663de122646a696b84999

mattsherman commented on packages/js/data/src/crud/resolvers.ts:
```diff
@@ -63,21 +67,69 @@ export const createResolvers = ( {
 
+	const getItemsTotalCount = function* ( query?: Partial< ItemQuery > ) {
+		const totalsQuery = {
+			...( typeof query === 'string' ? {} : query || {} ),
```
I don't understand two things about this `typeof query === 'string'` check (comment applies to other parts of the code that do a similar check as well)...

* Why are we checking the type of `query` at all, when we type `query` to be `Partial< ItemQuery >` in the params? Won't it always be `Partial< ItemQuery >` or `null`?
* If it does somehow get passed as a `string`, why do we just set things to an empty object?

louwie17 replied:
Yeah, it confused me at first to, and logic does indeed make it a bit more confusing.
It has to do with this [`applyNamespace` function](https://github.com/woocommerce/woocommerce/blob/trunk/packages/js/data/src/crud/utils.ts#L91-L98) that at-least the selectors get wrapped with, I should actually double check if the same thing happens for resolvers, otherwise the above won't be necessary.
The `applyNamespace` function tags on the `namespace` prop at the end of each function call.
This could mean if a user were to call `getItemsTotalCount()`, so without any parameters, the first parameter will be the `namespace` (which is a string). 
Although I only see the namespace being used in the `getItem` selectors so maybe the `applyNamespace` wrapper can be removed from all the other selectors -> https://github.com/woocommerce/woocommerce/blob/trunk/packages/js/data/src/crud/selectors.ts#L146-L162

cc: @joshuatf as he wrote this I believe.

louwie17 replied:
I end up adding a `defaultArgs` option to the `applyNamespace`, so this way the last param isn't always overwritten by the namespace. See here: https://github.com/woocommerce/woocommerce/pull/34613/commits/09ef544211684339f8976e604fdb61829d276c3e

mattsherman replied:
I honestly still find this whole logic super confusing.

louwie17 replied:
Yeah, it is a tad confusing I do agree.
The main problem this was set out to solve is to allow the crud store to support REST endpoints that rely on a parent.
Attribute terms is a good example that lives under this url: `products/attributes/<attribute_id>/terms/<id>` and I believe the the way it's stored in Redux is by prefixing the attribute_id before the term id.

Anyway it could be worth revisiting this logic at some point, as we could probably do something similar as what the resolvers do, by the use of the `createResolvers` function, so the `namespace` prop is more globally available instead of needing to be passed in through the use of `applyNamespace`.

mattsherman commented on packages/js/data/src/crud/selectors.ts:
```diff
@@ -113,6 +113,20 @@ export const getItems = createSelector(
 
+export const getItemsTotalCount = (
+	state: ResourceState,
+	query: ItemQuery | string
```
Why can query be an `ItemQuery` or a `string` here?

louwie17 replied:
See response in the previous comment: https://github.com/woocommerce/woocommerce/pull/34613#discussion_r966385117 which explains it.

louwie17 replied:
I end up removing the `| string` as I added the `defaultArgs` support.

mattsherman commented on packages/js/data/src/utils.ts:
```diff
@@ -21,6 +21,25 @@ export function getResourceName(
+ */
+export function getTotalCountResourceName(
+	prefix: string,
+	query: Record< string, unknown >
```
Why type `query` as `Record< string, unknown >` here instead of `Partial< ItemQuery >`?

louwie17 replied:
`ItemQuery` is unique to the `crud` datastore, given this util is in the main data package. `Record< string, unknown >` is more generic. I followed the same format as `getResourceName` above.

mattsherman reviewed (commented):
Thanks for the changes, @louwie17 !

I left some comments about parts of the code that I don't really understand. I'm not Approving or Requesting Changes at this point because that depends on a better understanding of those sections! 😄

louwie17 committed:
Remove deprecated data-controls function

louwie17 committed:
Add defaultArgs option for applyNamespace

louwie17 commented:
@mattsherman I addressed some of your feedback and also left replies to your comments with extra context and some of the reasons why.
This should be ready for another look.

mattsherman reviewed (approved):
Tests well. Nice work!

The only remaining issue I have is one of lingering confusion surrounding the `applyNamespace` logic and implementation. It feels more confusing than it should be. But, as this is existing low level functionality that just used by the functionality in this PR, I think we can merge this PR.

louwie17 commented:
> The only remaining issue I have is one of lingering confusion surrounding the applyNamespace logic and implementation. It feels more confusing than it should be. But, as this is existing low level functionality that just used by the functionality in this PR, I think we can merge this PR.

I do agree that it might be a bit to confusing, I left a more detailed reply in the comment above. Definitely worth revisiting.
